M2/Macaulay2/packages/A1BrouwerDegrees/Code/BuildingForms.m2

```diff
@@ -2,58 +2,152 @@
+isCompatibleElement (Ring, RingElement) := Boolean => (R, b) -> (
+    -- If the ring is the complex numbers, the real numbers, or the rational numbers, the scalar is not compatible. (The case when the scalar is a complex number, real number, rational number, or integer is treated in the previous variant.)
+    if instance(R, ComplexField) or instance(R, RealField) or R === QQ then return false;
+    -- If the ring is a finite field, require the scalar to be an element of that finite field. (The case when the scalar is a rational number or integer is treated in the previous variant.)
+    if instance(R, GaloisField) then (
+        if not (instance(ring b, GaloisField) and R.order == (ring b).order) then return false;
+    )
+    -- For other rings, require the scalar to be an element of the ring
+    else (
+        if ring b =!= R then return false;
+    );
+    -- If we get through the above, then the scalar is compatible with the ring
+    true
+)
+
+-- The third version of this function treats the case when R is an InexactFieldFamily and b is a Number (e.g. an element of RR_53, CC_53, QQ, or ZZ)
+isCompatibleElement (InexactFieldFamily, Number) := Boolean => (R, b) -> (
+    -- If the inexact field family is the complex numbers, require the scalar to be a complex number, a real number, a rational number, or an integer
+    if R === CC then (
+        if not (instance(ring b, ComplexField) or instance(ring b, RealField) or ring b === QQ or ring b === ZZ) then return false;
     )
+    -- If the inexact field family is the real numbers, require the scalar to be a real number, a rational number, or an integer
+    else if R === RR then (
+        if not (instance(ring b, RealField) or ring b === QQ or ring b === ZZ) then return false;
+    )
+    -- For any other inexact field families, require the scalar to be an element of the inexact field family
+    else (
+        if ring b =!= R then return false;
+    );
+    -- If we get through the above, then the scalar is compatible with the ring
+    true
+)
+
+-- The fourth version of this function treats the case when R is an InexactFieldFamily and b is a RingElement (e.g. an element of a finite field)
+isCompatibleElement (InexactFieldFamily, RingElement) := Boolean => (R, b) -> (
+    -- Here we require the scalar to be an element of the inexact field family
+    ring b === R
+)
 
-makeDiagonalForm (Ring, ZZ) := GrothendieckWittClass => (kk, a) -> (
-    makeGWClass matrix(kk, {{sub(a, kk)}})
+-- Input: A field kk of characteristic not 2, and a list of elements b_1,...,b_n of kk
+-- Output: The Grothendieck-Witt class represented by the diagonal form <b_1,...,b_n> 
+
+makeDiagonalForm = method()
+makeDiagonalForm (Ring, RingElement) := GrothendieckWittClass => (kk, b) -> (
+    if not isCompatibleElement(kk, b) then error "scalar not compatible with field";
+    makeGWClass matrix(kk, {{substitute(b, kk)}})
```

Comment:
`isCompatibleElement` seems like a lot of work to make sure that `b` can be promoted to `kk`, but all that checking will be done already by matrix.  Why not just `matrix(kk, {{b}})`?

Reply:
I can't seem to find the source code for this in M2, but does it rely on `substitute` or `promote`? Because we were originally intending to work with these existing methods, but they don't always work the way we expect.

Reply:
The code's a bit hidden (`matrix(Ring, List)` calls `map(Module, Nothing, List)`, which calls the local helper function `makeRawTable`), but in the end, it does call `promote` on all the entries.